R/dtc_create_iso8601.R

```diff
@@ -0,0 +1,415 @@
+#'
+#' # Two-digit years are supported
+#' create_iso8601(c("20-01-01", "200101"), .format = list(c("y-m-d", "ymd")))
+#'
+#' # `.cutoff_2000` sets the cutoff for two-digit to four-digit year conversion
+#' # Default is at 68.
+#' create_iso8601(c("67-01-01", "68-01-01", "69-01-01"), .format = "y-m-d")
+#'
+#' # Change it to 80.
+#' create_iso8601(c("79-01-01", "80-01-01", "81-01-01"), .format = "y-m-d", .cutoff_2000 = 80)
+#'
+#' # Converting times
+#' create_iso8601("15:10", .format = "HH:MM", .check_format = FALSE)
+#' create_iso8601("2:10", .format = "HH:MM", .check_format = FALSE)
+#' create_iso8601("2:1", .format = "HH:MM", .check_format = FALSE)
+#' create_iso8601("02:01:56", .format = "HH:MM:SS", .check_format = FALSE)
+#' create_iso8601("020156.5", .format = "HHMMSS", .check_format = FALSE)
+#'
+#' # Converting date-times
+#' create_iso8601("12 NOV 202015:15", .format = "dd mmm yyyyHH:MM", .check_format = FALSE)
+#'
+#' # Indicate allowed missing values to make the parsing pass
+#' create_iso8601("U DEC 201914:00", .format = "dd mmm yyyyHH:MM", .check_format = FALSE)
+#' create_iso8601("U DEC 201914:00", .format = "dd mmm yyyyHH:MM", .check_format = FALSE, .na = "U")
+#'
+#' create_iso8601("NOV 2020", .format = "m y", .check_format = FALSE)
+#' create_iso8601(c("MAR 2019", "MaR 2020", "mar 2021"), .format = "m y", .check_format = FALSE)
+#'
+#' create_iso8601("2019-04-041045-", .format = "yyyy-mm-ddHHMM-", .check_format = FALSE)
+#'
+#' create_iso8601("20200507null", .format = "ymd(HH:MM:SS)", .check_format = FALSE)
+#' create_iso8601("20200507null", .format = "ymd((HH:MM:SS)|null)", .check_format = FALSE)
+#'
+#' # Fractional seconds
+#' create_iso8601("2019-120602:20:13.1230001", .format = "y-mdH:M:S", .check_format = FALSE)
+#' @export
+create_iso8601 <- function(..., .format, .fmt_c = fmt_cmp(), .na = NULL, .cutoff_2000 = 68L, .check_format = FALSE) {
+
+  assert_fmt_c(.fmt_c)
+
+  dots <- rlang::dots_list(...)
+
+  if (rlang::is_empty(dots)) {
+    return(character())
+  }
+
+  # Check if all vectors in `dots` are of character type.
+  if (!identical(unique(sapply(dots, typeof)), "character")) {
+    rlang::abort("All vectors in `...` must be of type character.")
+  }
+
+  # Check if all vectors in `dots` are of the same length.
+  n <- unique(lengths(dots))
+  if (!identical(length(n), 1L)) {
+    rlang::abort("All vectors in `...` must be of the same length.")
+  }
+
+  if (!identical(length(dots), length(.format))) {
+    rlang::abort("Number of vectors in `...` should match length of `.format`.")
+  }
+
+  # Check that the `.format` is either a character vector or a list of
+  # character vectors, and that each string is one of the possible formats.
+  if (.check_format) assert_dtc_format(.format)
+
+  cap_matrices <- purrr::map2(dots, .format, ~ parse_dttm(dttm = .x, fmt = .y, na = .na, fmt_c = .fmt_c))
+  cap_matrix <- coalesce_capture_matrices(!!!cap_matrices)
+
+  format_iso8601(cap_matrix, .cutoff_2000 = .cutoff_2000)
```

Comment:
Didn't we agree to pass a warning message if some values do not adhere to the specified format?

create_iso8601(c("UN UNK 201914:00", "UN JAN 2021"), .format = "dd mmm yyyy", .check_format = FALSE, .na = c("UN", "UNK"))

[1] NA        "2021-01"

The warning message can read as "Some source values does not adhere to the format provided, returning null values"

Maybe we need to create an exception listing. It can be developed as a separate issue.

Reply:
Yes, we did agree on this. But like you said, I think it is best provided as a separate issue.

Reply:
For the warnings' feature and feedback on which values failed, let us track the progress here: https://github.com/pharmaverse/sdtm.oak/issues/29.